VMDK plugin returns confusing message

When VMDK plugin is being restarted (e.g. due to VM restart), the background refcnt thread needs some time to initialize. During the initialization process, the plugin might get incoming mount/unmount/remove volume requests. These operations will need to get the ref counts of the volume. In this case , the current logic will return a hard-coded value 1, indicating the refcnt is still in initialization process. This caused very confusing error message being returned to the caller. For example, after verifying the volume status is detached, removing this volume still failed with error message: "Remove failure - volume is still mounted", which is very confusing and misleading.

As we discussed, we should be using a special value, such as -1, to indicate this special status when refcount is still initializing. Using 1 will make it confusing with the normal scenario where the volume is truly being referenced by one container.

We need to fix all 3 functions - vmdk.getRefCount(), vmdk.incrRefCount() and vmdk.decrRefCount() - to return this special value, and fixing the calling APIs accordingly to check this special value and handle the logic appropriately.

This issue has caused vmlistener test case fails intermittently - see issue #1308

Here's the stack trace of the vmlistener test case failure:

2017/06/04 07:26:41 Failed to invoke command [docker volume rm vmlistener_test_volume_1496561145]: exit status 1

----------------------------------------------------------------------
FAIL: vmlistener_test.go:61: VMListenerTestParams.TearDownTest

vmlistener_test.go:67:
    c.Assert(err, IsNil, Commentf(out))
... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc42017ce00), Stderr:[]uint8(nil)} ("exit status 1")
... Error response from daemon: unable to remove volume: remove vmlistener_test_volume_1496561145: VolumeDriver.Remove: Remove failure - volume is still mounted.  volume=vmlistener_test_volume_1496561145, refcount=1


----------------------------------------------------------------------
PANIC: vmlistener_test.go:80: VMListenerTestParams.TestKillVM

... Panic: Fixture has panicked (see related PANIC)
--- FAIL: Test (351.59s)

--- a/vmdk_plugin/drivers/photon/photon_driver.go
+++ b/vmdk_plugin/drivers/photon/photon_driver.go
@@ -114,7 +114,7 @@ func NewVolumeDriver(targetURL string, projectID string, hostID string, mountDir
 
 // Return the number of references for the given volume
 func (d *VolumeDriver) getRefCount(vol string) uint {
-	if d.refCounts.GetInitSuccess() != true {
+	if d.refCounts.IsInitialized() != true {
 		return 1
 	}
 	return d.refCounts.GetCount(vol)
@@ -122,7 +122,7 @@ func (d *VolumeDriver) getRefCount(vol string) uint {
 
 // Increment the reference count for the given volume
 func (d *VolumeDriver) incrRefCount(vol string) uint {
-	if d.refCounts.GetInitSuccess() != true {
+	if d.refCounts.IsInitialized() != true {
 		return 1
 	}
 	return d.refCounts.Incr(vol)
@@ -130,7 +130,7 @@ func (d *VolumeDriver) incrRefCount(vol string) uint {
 
 // Decrement the reference count for the given volume
 func (d *VolumeDriver) decrRefCount(vol string) (uint, error) {
-	if d.refCounts.GetInitSuccess() != true {
+	if d.refCounts.IsInitialized() != true {
 		return 1, nil
 	}
 	return d.refCounts.Decr(vol)
@@ -560,8 +560,10 @@ func (d *VolumeDriver) Create(r volume.Request) volume.Response {
 func (d *VolumeDriver) Remove(r volume.Request) volume.Response {
 	log.WithFields(log.Fields{"name": r.Name}).Info("Removing volume ")
 
-	if d.refCounts.GetInitSuccess() != true {
-		msg := fmt.Sprintf(plugin_utils.ErrorPluginInit+" Cannot remove volume=%s", r.Name)
+	// Cannot remove volumes till plugin completely initializes (refcounting is complete)
+	// because we don't know if it is being used or not
+	if d.refCounts.IsInitialized() != true {
+		msg := fmt.Sprintf(plugin_utils.PluginInitError+" Cannot remove volume=%s", r.Name)
 		log.Error(msg)
 		return volume.Response{Err: msg}
 	}
@@ -639,7 +641,7 @@ func (d *VolumeDriver) Unmount(r volume.UnmountRequest) volume.Response {
 	d.refCounts.StateMtx.Lock()
 	defer d.refCounts.StateMtx.Unlock()
 
-	if d.refCounts.GetInitSuccess() != true {
+	if d.refCounts.IsInitialized() != true {
 		// if refcounting hasn't been succesful,
 		// no refcounting, no unmount. All unmounts are delayed
 		// until we succesfully populate the refcount map

--- a/vmdk_plugin/drivers/photon/photon_driver.go
+++ b/vmdk_plugin/drivers/photon/photon_driver.go
@@ -560,7 +560,13 @@ func (d *VolumeDriver) Create(r volume.Request) volume.Response {
 func (d *VolumeDriver) Remove(r volume.Request) volume.Response {
 	log.WithFields(log.Fields{"name": r.Name}).Info("Removing volume ")
 
-	// Docker is supposed to block 'remove' command if the volume is used. Verify.
+	if d.refCounts.GetInitSuccess() != true {
+		msg := fmt.Sprintf(plugin_utils.ErrorPluginInit+" Cannot remove volume=%s", r.Name)
+		log.Error(msg)
+		return volume.Response{Err: msg}
+	}
+
+	// Docker is supposed to block 'remove' command if the volume is used.
 	if d.getRefCount(r.Name) != 0 {
 		msg := fmt.Sprintf("Remove failure - volume is still mounted. "+
 			" volume=%s, refcount=%d", r.Name, d.getRefCount(r.Name))

--- a/vmdk_plugin/drivers/vmdk/vmdk_driver.go
+++ b/vmdk_plugin/drivers/vmdk/vmdk_driver.go
@@ -422,7 +422,13 @@ func (d *VolumeDriver) Create(r volume.Request) volume.Response {
 func (d *VolumeDriver) Remove(r volume.Request) volume.Response {
 	log.WithFields(log.Fields{"name": r.Name}).Info("Removing volume ")
 
-	// Docker is supposed to block 'remove' command if the volume is used. Verify.
+	if d.refCounts.GetInitSuccess() != true {
+		msg := fmt.Sprintf(plugin_utils.ErrorPluginInit+" Cannot remove volume=%s", r.Name)
+		log.Error(msg)
+		return volume.Response{Err: msg}
+	}
+
+	// Docker is supposed to block 'remove' command if the volume is used.
 	if d.getRefCount(r.Name) != 0 {
 		msg := fmt.Sprintf("Remove failure - volume is still mounted. "+
 			" volume=%s, refcount=%d", r.Name, d.getRefCount(r.Name))

--- a/vmdk_plugin/drivers/vmdk/vmdk_driver.go
+++ b/vmdk_plugin/drivers/vmdk/vmdk_driver.go
@@ -103,7 +103,7 @@ func (d *VolumeDriver) VolumesInRefMap() []string {
 
 // Return the number of references for the given volume
 func (d *VolumeDriver) getRefCount(vol string) uint {
-	if d.refCounts.GetInitSuccess() != true {
+	if d.refCounts.IsInitialized() != true {
 		return 1
 	}
 	return d.refCounts.GetCount(vol)
@@ -111,7 +111,7 @@ func (d *VolumeDriver) getRefCount(vol string) uint {
 
 // Increment the reference count for the given volume
 func (d *VolumeDriver) incrRefCount(vol string) uint {
-	if d.refCounts.GetInitSuccess() != true {
+	if d.refCounts.IsInitialized() != true {
 		return 1
 	}
 	return d.refCounts.Incr(vol)
@@ -119,7 +119,7 @@ func (d *VolumeDriver) incrRefCount(vol string) uint {
 
 // Decrement the reference count for the given volume
 func (d *VolumeDriver) decrRefCount(vol string) (uint, error) {
-	if d.refCounts.GetInitSuccess() != true {
+	if d.refCounts.IsInitialized() != true {
 		return 1, nil
 	}
 	return d.refCounts.Decr(vol)
@@ -422,8 +422,10 @@ func (d *VolumeDriver) Create(r volume.Request) volume.Response {
 func (d *VolumeDriver) Remove(r volume.Request) volume.Response {
 	log.WithFields(log.Fields{"name": r.Name}).Info("Removing volume ")
 
-	if d.refCounts.GetInitSuccess() != true {
-		msg := fmt.Sprintf(plugin_utils.ErrorPluginInit+" Cannot remove volume=%s", r.Name)
+	// Cannot remove volumes till plugin completely initializes (refcounting is complete)
+	// because we don't know if it is being used or not
+	if d.refCounts.IsInitialized() != true {
+		msg := fmt.Sprintf(plugin_utils.PluginInitError+" Cannot remove volume=%s", r.Name)
 		log.Error(msg)
 		return volume.Response{Err: msg}
 	}
@@ -483,7 +485,7 @@ func (d *VolumeDriver) Unmount(r volume.UnmountRequest) volume.Response {
 	d.refCounts.StateMtx.Lock()
 	defer d.refCounts.StateMtx.Unlock()
 
-	if d.refCounts.GetInitSuccess() != true {
+	if d.refCounts.IsInitialized() != true {
 		// if refcounting hasn't been succesful,
 		// no refcounting, no unmount. All unmounts are delayed
 		// until we succesfully populate the refcount map

--- a/vmdk_plugin/utils/plugin_utils/plugin_utils.go
+++ b/vmdk_plugin/utils/plugin_utils/plugin_utils.go
@@ -32,6 +32,9 @@ const (
 	// index datastore from volume meta
 	// "datastore" key is defined in vmdkops service
 	datastoreKey = "datastore"
+
+	// ErrorPluginInit message to indicate that plugin initialization(refcounting) is not yet complete
+	ErrorPluginInit = "Plugin initialization in progress."
 )
 
 // VolumeInfo - Volume fullname, datastore and metadata

--- a/vmdk_plugin/utils/plugin_utils/plugin_utils.go
+++ b/vmdk_plugin/utils/plugin_utils/plugin_utils.go
@@ -33,8 +33,8 @@ const (
 	// "datastore" key is defined in vmdkops service
 	datastoreKey = "datastore"
 
-	// ErrorPluginInit message to indicate that plugin initialization(refcounting) is not yet complete
-	ErrorPluginInit = "Plugin initialization in progress."
+	// PluginInitError message to indicate that plugin initialization(refcounting) is not yet complete
+	PluginInitError = "Plugin initialization in progress."
 )
 
 // VolumeInfo - Volume fullname, datastore and metadata

--- a/vmdk_plugin/utils/refcount/refcnt.go
+++ b/vmdk_plugin/utils/refcount/refcnt.go
@@ -158,7 +158,7 @@ func newRefCount() *refCount {
 }
 
 // return if refcount initialization has been successful
-func (r *RefCountsMap) GetInitSuccess() bool {
+func (r *RefCountsMap) IsInitialized() bool {
 	return r.refcntInitSuccess
 }
 

